packages/grafana-ui/src/components/Dropdown/Dropdown.tsx

```diff
@@ -36,20 +36,26 @@ export const Dropdown = React.memo(({ children, overlay, placement }: Props) =>
     setShow(false);
   };
 
+  const handleKeys = (event: React.KeyboardEvent) => {
+    if (event.key === 'Escape') {
+      setShow(false);
+    }
+  };
+
   return (
     <>
       {React.cloneElement(typeof children === 'function' ? children(visible) : children, {
         ref: setTriggerRef,
       })}
       {visible && (
         <Portal>
-          <FocusScope autoFocus>
+          <FocusScope autoFocus restoreFocus contain>
```

Comment:
I think the addition of `contain` here is not good... Because it means that pressing 'Tab' doesn't do anything, when it should allow the user to move on to the next element. Maybe the best solution might be to move this to use a `useDialog` like we have in other places. Wdyt?

Reply:
so there's a couple of options. apologies in advance for brain dump... 😅 

the other places we do similar stuff, we use `contain` and 'Tab' just cycles through the focusable elements in that overlay. e.g. `TimeRangePicker`. but that's for a generic overlay and this is a dropdown, so maybe the behaviour should be slightly different. part of the reason 'Tab' doesn't do anything in **this** case (when we use `Menu`) is because of the custom key logic in `Menu`: https://github.com/grafana/grafana/blob/ash/dropdown-escape/packages/grafana-ui/src/components/Menu/hooks.ts#L103. we could change that to focus the next element (as e.g. 'ArrowDown' does), but looking at other `Dropdown`s it seems like 'Tab' first closes the overlay before moving focus again.

separately, we could also use `useOverlay` + `useDialog`. it ends up a little more complicated just because of the shared refs between those hooks and the `usePopperTooltip` hook. we end up with something like:

```typescript
  const { getArrowProps, getTooltipProps, setTooltipRef, setTriggerRef, visible } = usePopperTooltip({
    visible: show,
    placement: placement,
    onVisibleChange: setShow,
    interactive: true,
    delayHide: 0,
    delayShow: 0,
    offset: [0, 8],
    trigger: ['click'],
  });
  const ref: MutableRefObject<HTMLDivElement | null> = useRef<HTMLDivElement>(null);
  const handleRef: React.RefCallback<HTMLDivElement> = (domRef) => {
    setTooltipRef(domRef);
    ref.current = domRef;
  }
  const { overlayProps } = useOverlay({ onClose: () => setShow(false), isDismissable: true, isOpen: show }, ref);
  const { dialogProps } = useDialog({}, ref);
```

this means we can remove the `handleKeys` method, but it still leaves the problem of what 'Tab' should do. if we remove `contain`, 'Tab' moves to the next item in the DOM outside of the menu but **does not** close the menu. that's slightly better than today where the 'Tab' moves to the next element outside of the menu **and** your tabbing context is moved to the very end 🤮 but not much better

having written this all down i think the best option here is to not use `useOverlay` and `useDialog`, keep `contain` and make 'Tab' do the same thing as 'Escape' in the `handleKeys` method. that way, pressing 'Tab' whilst the menu is open will close the menu and restore focus to the dropdown trigger, then you can continue tabbing away. wdyt? 🤔

Reply:
I think this use of Tab is better than what is there right now, so even if it's not perfect it's an improvement 👏